ledger/participant-integration-api/src/main/scala/platform/index/IncrementalIndexPruningService.scala

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.index
+
+import com.daml.ledger.offset.Offset
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.store.dao.LedgerReadDao
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class IncrementalIndexPruningService(ledgerReadDao: LedgerReadDao, maximumPruningWindowSize: Int) {
+  private val logger = ContextualizedLogger.get(getClass)
+
+  def prune(pruneUpToInclusive: Offset, pruneAllDivulgedContracts: Boolean)(implicit
+      loggingContext: LoggingContext
+  ): Future[Unit] = {
+    implicit val ec: ExecutionContext = ExecutionContext.parasitic
+    for {
+      pruningOffsets <- ledgerReadDao.pruningOffsets
+      startPruningOffset = getStartPruningOffset(pruneAllDivulgedContracts, pruningOffsets)
+
+      _ <- pruneIncrementally(
+        latestPrunedUpToInclusive = startPruningOffset,
+        pruneUpToInclusive = pruneUpToInclusive,
+        pruneAllDivulgedContracts = pruneAllDivulgedContracts,
+        maxPruningWindowSize = maximumPruningWindowSize,
+      )
+    } yield ()
+  }
+
+  private def pruneIncrementally(
+      latestPrunedUpToInclusive: Offset,
+      pruneUpToInclusive: Offset,
+      pruneAllDivulgedContracts: Boolean,
+      maxPruningWindowSize: Int,
+  )(implicit
+      loggingContext: LoggingContext,
+      executionContext: ExecutionContext,
+  ): Future[Unit] = {
+    def go(pruningWindowStartExclusive: Offset): Future[Unit] =
+      // If start exclusive is not smaller than the requested `pruneUpToInclusive`, we're done
+      if (pruningWindowStartExclusive < pruneUpToInclusive)
```

Comment:
how do we stop this?
* we need to make sure, that we know what happens with an ongoing huge pruning if node is passivated, node is being stopped etc
* we need to make findings explicit, and document operational aspects to help operators